FIAT/kong_mulder_veldhuizen.py

```diff
@@ -0,0 +1,165 @@
+                1: dict((i, [i + 3]) for i in range(3)),
+                2: {0: [6]},
+            }
+        elif sd == 3:
+            entity_ids = {
+                0: dict((i, [i]) for i in range(4)),
+                1: dict((i, [i + 4]) for i in range(6)),
+                2: dict((i, [i + 10]) for i in range(4)),
+                3: {0: [14]},
+            }
+    elif degree == 3:
+        if sd == 2:
+            etop = [[3, 4], [6, 5], [7, 8]]
+            entity_ids = {
+                0: dict((i, [i]) for i in range(3)),
+                1: dict((i, etop[i]) for i in range(3)),
+                2: {0: [9, 10, 11]},
+            }
+        elif sd == 3:
+            etop = [[4, 5], [7, 6], [8, 9], [11, 10], [12, 13], [14, 15]]
+            ftop = [[16, 17, 18], [19, 20, 21], [22, 23, 24], [25, 26, 27]]
+            entity_ids = {
+                0: dict((i, [i]) for i in range(4)),
+                1: dict((i, etop[i]) for i in range(6)),
+                2: dict((i, ftop[i]) for i in range(4)),
+                3: {0: [28, 29, 30, 31]},
+            }
+    elif degree == 4:
+        if sd == 2:
+            etop = [[6, 3, 7], [9, 4, 8], [10, 5, 11]]
+            entity_ids = {
+                0: dict((i, [i]) for i in range(3)),
+                1: dict((i, etop[i]) for i in range(3)),
+                2: {0: [i for i in range(12, 18)]},
+            }
+    elif degree == 5:
+        if sd == 2:
+            etop = [[9, 3, 4, 10], [12, 6, 5, 11], [13, 7, 8, 14]]
+            entity_ids = {
+                0: dict((i, [i]) for i in range(3)),
+                1: dict((i, etop[i]) for i in range(3)),
+                2: {0: [i for i in range(15, 30)]},
+            }
+    return entity_ids
+
+
+def bump(T, deg):
+    """Increase degree of polynomial along"""
+    sd = T.get_spatial_dimension()
+    if deg == 1:
+        return (0,)
+    if sd == 2:
+        if deg < 5:
+            return (1,)
+        elif deg == 5:
+            return (2,)
+    elif sd == 3:
+        return (1, 2)
+
+
+def KongMulderVeldhuizenSpace(T, deg):
+    sd = T.get_spatial_dimension()
+    if deg == 1:
+        return Lagrange(T, 1).poly_set
+    else:
+        L = Lagrange(T, deg)
+        # Toss the bubble from Lagrange since it's dependent
+        # on the higher-dimensional bubbles
+        if sd == 2:
+            inds = [
+                i
+                for i in range(L.space_dimension())
+                if i not in L.dual.entity_ids[sd][0]
+            ]
+        elif sd == 3:
+            not_inds = [L.dual.entity_ids[sd][0]] + [
+                L.dual.entity_ids[sd - 1][f] for f in L.dual.entity_ids[sd - 1]
+            ]
+            not_inds = [item for sublist in not_inds for item in sublist]
+            inds = [i for i in range(L.space_dimension()) if i not in not_inds]
+        RL = RestrictedElement(L, inds)
+        # interior cell bubble
+        bubs = Bubble(T, deg + max(bump(T, deg)))
```

Comment:
Is this logic right?  In 3d, you get one degree for edges and one for faces.  Shouldn't you just be using entries [0] and [1] in that case?

Reply:
It produces the correct space for triangles as well, yes. The `max`/`min` of a 1 element list is the value of the element so this allows us to consider both cases  (2 and 3d) with the same code.